File: kaleido-gradle-plugin/src/main/java/com/tongsr/kaleido/gradle/ApplicationResourceInventory.java
package com.tongsr.kaleido.gradle;

import java.io.IOException;
import java.nio.file.Files;
import java.nio.file.Path;
import java.util.Comparator;
import java.util.List;
import java.util.Set;
import java.util.TreeSet;
import java.util.regex.Pattern;
import javax.xml.XMLConstants;
import javax.xml.parsers.DocumentBuilderFactory;
import org.w3c.dom.Element;

final class ApplicationResourceInventory {
    private ApplicationResourceInventory() {}

    static Result scan(Iterable<Path> roots, String project, String variant) throws IOException {
        var resources = new TreeSet<BundleRewriteModule.ResourceKey>(Comparator
                .comparing(BundleRewriteModule.ResourceKey::type)
                .thenComparing(BundleRewriteModule.ResourceKey::name));
        var publicNames = new TreeSet<String>();
        var keepSelectors = new TreeSet<String>();
        for (var root : roots) {
            if (!Files.isDirectory(root)) continue;
            try (var paths = Files.walk(root)) {
                for (var file : paths.filter(Files::isRegularFile).sorted().toList()) {
                    var relative = root.relativize(file);
                    if (relative.getNameCount() < 2) continue;
                    var directory = relative.getName(0).toString();
                    var type = directory.split("-", 2)[0];
                    if (file.getFileName().toString().endsWith(".xml")) {
                        scanToolsKeep(file, keepSelectors, project, variant);
                    }
                    if ("values".equals(type) && file.getFileName().toString().endsWith(".xml")) {
                        scanValues(file, resources, publicNames, project, variant);
                    } else {
                        var name = resourceFileName(file.getFileName().toString());
                        if (!name.isBlank()) resources.add(
                                new BundleRewriteModule.ResourceKey(type, name));
                    }
                }
            }
        }
        var protectedNames = new TreeSet<>(publicNames);
        for (var resource : resources) {
            var identity = resource.type() + "/" + resource.name();
            if (keepSelectors.stream().anyMatch(
                    selector -> glob(selector).matcher(identity).matches())) {
                protectedNames.add(resource.name());
            }
        }
        var warnings = keepSelectors.stream()
                .filter(selector -> selector.contains("*") || selector.contains("?"))
                .map(selector -> "KLD-RESOURCE-001 project=" + project + " variant=" + variant
                        + " stage=bundle-rewrite origin=tools:keep target=" + selector
                        + " reason=Broad tools:keep pattern retained matching resources"
                        + " repair=<none>")
                .toList();
        return new Result(Set.copyOf(resources), Set.copyOf(protectedNames), warnings);
    }

    private static void scanValues(
            Path file,
            Set<BundleRewriteModule.ResourceKey> resources,
            Set<String> publicNames,
            String project,
            String variant) {
        try {
            var factory = secureFactory();
            var document = factory.newDocumentBuilder().parse(file.toFile());
            var root = document.getDocumentElement();
            for (var index = 0; index < root.getChildNodes().getLength(); index++) {
                if (!(root.getChildNodes().item(index) instanceof Element element)) continue;
                var tag = element.getTagName();
                var name = element.getAttribute("name");
                var type = valuesType(tag, element.getAttribute("type"));
                if (!name.isBlank() && type != null) {
                    resources.add(new BundleRewriteModule.ResourceKey(type, name));
                    if ("public".equals(tag)) publicNames.add(name);
                }
                if ("declare-styleable".equals(tag)) {
                    var attrs = element.getElementsByTagName("attr");
                    for (var attrIndex = 0; attrIndex < attrs.getLength(); attrIndex++) {
                        var attr = (Element) attrs.item(attrIndex);
                        var attrName = attr.getAttribute("name");
                        if (!attrName.isBlank() && !attrName.startsWith("android:")) {
                            resources.add(new BundleRewriteModule.ResourceKey("attr", attrName));
                        }
                    }
                }
            }
        } catch (Exception invalid) {
            throw new KaleidoDiagnostic("KLD-BUNDLE-001", project, variant,
                    "bundle-rewrite", BundleRewriteArtifacts.PLAN_SCHEMA, file.toString(),
                    "Application values XML cannot be inventoried safely",
                    "Fix the malformed values resource before Release").failure();
        }
    }

    private static void scanToolsKeep(
            Path file, Set<String> selectors, String project, String variant) {
        try {
            var document = secureFactory().newDocumentBuilder().parse(file.toFile());
            var elements = document.getElementsByTagName("*");
            for (var index = 0; index < elements.getLength(); index++) {
                var element = (Element) elements.item(index);
                var keep = element.getAttributeNS("http://schemas.android.com/tools", "keep");
                if (keep.isBlank()) continue;
                for (var value : keep.split(",")) {
                    var normalized = value.trim();
                    if (normalized.startsWith("@")) normalized = normalized.substring(1);
                    var colon = normalized.indexOf(':');
                    if (colon >= 0) normalized = normalized.substring(colon + 1);
                    if (normalized.matches("[a-z*][a-z0-9_*?]*/[a-z0-9_*?]+")) {
                        selectors.add(normalized);
                    }
                }
            }
        } catch (Exception invalid) {
            throw new KaleidoDiagnostic("KLD-BUNDLE-001", project, variant,
                    "bundle-rewrite", BundleRewriteArtifacts.PLAN_SCHEMA, file.toString(),
                    "Application XML protection attributes cannot be inventoried safely",
                    "Fix the malformed resource XML before Release").failure();
        }
    }

    private static DocumentBuilderFactory secureFactory() throws Exception {
        var factory = DocumentBuilderFactory.newInstance();
        factory.setNamespaceAware(true);
        factory.setFeature("http://apache.org/xml/features/disallow-doctype-decl", true);
        factory.setFeature("http://xml.org/sax/features/external-general-entities", false);
        factory.setFeature("http://xml.org/sax/features/external-parameter-entities", false);
        factory.setAttribute(XMLConstants.ACCESS_EXTERNAL_DTD, "");
        factory.setAttribute(XMLConstants.ACCESS_EXTERNAL_SCHEMA, "");
        return factory;
    }

    private static Pattern glob(String selector) {
        var regex = new StringBuilder("^");
        for (var character : selector.toCharArray()) {
            if (character == '*') regex.append(".*");
            else if (character == '?') regex.append('.');
            else regex.append(Pattern.quote(Character.toString(character)));
        }
        return Pattern.compile(regex.append('$').toString());
    }

    private static String valuesType(String tag, String declaredType) {
        if ("item".equals(tag) || "public".equals(tag)) {
            return declaredType.isBlank() ? null : declaredType;
        }
        if ("string-array".equals(tag) || "integer-array".equals(tag)) return "array";
        if ("declare-styleable".equals(tag)) return "styleable";
        if (Set.of("eat-comment", "skip", "overlayable").contains(tag)) return null;
        return tag;
    }

    private static String resourceFileName(String fileName) {
        var ninePatch = fileName.indexOf(".9.");
        if (ninePatch > 0) return fileName.substring(0, ninePatch);
        var dot = fileName.indexOf('.');
        return dot > 0 ? fileName.substring(0, dot) : fileName;
    }

    record Result(Set<BundleRewriteModule.ResourceKey> resources,
                  Set<String> protectedNames, List<String> warnings) {
        Result { warnings = warnings.stream().sorted().toList(); }
    }
}

File: kaleido-gradle-plugin/src/main/kotlin/com/tongsr/kaleido/gradle/ApplicationResourceInventory.kt
package com.tongsr.kaleido.gradle

import java.nio.file.Files
import java.nio.file.Path
import java.util.TreeSet
import java.util.regex.Pattern
import javax.xml.XMLConstants
import javax.xml.parsers.DocumentBuilderFactory
import org.w3c.dom.Element

internal object ApplicationResourceInventory {
    @JvmStatic
    fun scan(roots: Iterable<Path>, project: String, variant: String): Result {
        val resources = TreeSet(
            compareBy(BundleRewriteModule.ResourceKey::type)
                .thenBy(BundleRewriteModule.ResourceKey::name),
        )
        val publicNames = TreeSet<String>()
        val keepSelectors = TreeSet<String>()
        for (root in roots) {
            if (!Files.isDirectory(root)) continue
            Files.walk(root).use { paths ->
                for (file in paths.filter { Files.isRegularFile(it) }.sorted().toList()) {
                    val relative = root.relativize(file)
                    if (relative.nameCount < 2) continue
                    val directory = relative.getName(0).toString()
                    val type = directory.split("-", limit = 2)[0]
                    if (file.fileName.toString().endsWith(".xml")) {
                        scanToolsKeep(file, keepSelectors, project, variant)
                    }
                    if (type == "values" && file.fileName.toString().endsWith(".xml")) {
                        scanValues(file, resources, publicNames, project, variant)
                    } else {
                        val name = resourceFileName(file.fileName.toString())
                        if (name.isNotBlank()) {
                            resources.add(BundleRewriteModule.ResourceKey(type, name))
                        }
                    }
                }
            }
        }
        val protectedNames = TreeSet(publicNames)
        for (resource in resources) {
            val identity = resource.type + "/" + resource.name
            if (keepSelectors.any { glob(it).matcher(identity).matches() }) {
                protectedNames.add(resource.name)
            }
        }
        val warnings = keepSelectors
            .filter { selector -> selector.contains("*") || selector.contains("?") }
            .map { selector ->
                "KLD-RESOURCE-001 project=$project variant=$variant" +
                    " stage=bundle-rewrite origin=tools:keep target=$selector" +
                    " reason=Broad tools:keep pattern retained matching resources" +
                    " repair=<none>"
            }
        return Result(resources.toSet(), protectedNames.toSet(), warnings)
    }

    private fun scanValues(
        file: Path,
        resources: MutableSet<BundleRewriteModule.ResourceKey>,
        publicNames: MutableSet<String>,
        project: String,
        variant: String,
    ) {
        try {
            val factory = secureFactory()
            val document = factory.newDocumentBuilder().parse(file.toFile())
            val root = document.documentElement
            for (index in 0 until root.childNodes.length) {
                val node = root.childNodes.item(index)
                if (node !is Element) continue
                val tag = node.tagName
                val name = node.getAttribute("name")
                val type = valuesType(tag, node.getAttribute("type"))
                if (name.isNotBlank() && type != null) {
                    resources.add(BundleRewriteModule.ResourceKey(type, name))
                    if (tag == "public") publicNames.add(name)
                }
                if (tag == "declare-styleable") {
                    val attrs = node.getElementsByTagName("attr")
                    for (attrIndex in 0 until attrs.length) {
                        val attr = attrs.item(attrIndex) as Element
                        val attrName = attr.getAttribute("name")
                        if (attrName.isNotBlank() && !attrName.startsWith("android:")) {
                            resources.add(BundleRewriteModule.ResourceKey("attr", attrName))
                        }
                    }
                }
            }
        } catch (invalid: Exception) {
            throw KaleidoDiagnostic(
                "KLD-BUNDLE-001",
                project,
                variant,
                "bundle-rewrite",
                BundleRewriteArtifacts.PLAN_SCHEMA,
                file.toString(),
                "Application values XML cannot be inventoried safely",
                "Fix the malformed values resource before Release",
            ).failure()
        }
    }

    private fun scanToolsKeep(
        file: Path,
        selectors: MutableSet<String>,
        project: String,
        variant: String,
    ) {
        try {
            val document = secureFactory().newDocumentBuilder().parse(file.toFile())
            val elements = document.getElementsByTagName("*")
            for (index in 0 until elements.length) {
                val element = elements.item(index) as Element
                val keep = element.getAttributeNS("http://schemas.android.com/tools", "keep")
                if (keep.isBlank()) continue
                for (value in keep.split(",")) {
                    var normalized = value.trim()
                    if (normalized.startsWith("@")) normalized = normalized.substring(1)
                    val colon = normalized.indexOf(':')
                    if (colon >= 0) normalized = normalized.substring(colon + 1)
                    if (normalized.matches(Regex("[a-z*][a-z0-9_*?]*/[a-z0-9_*?]+"))) {
                        selectors.add(normalized)
                    }
                }
            }
        } catch (invalid: Exception) {
            throw KaleidoDiagnostic(
                "KLD-BUNDLE-001",
                project,
                variant,
                "bundle-rewrite",
                BundleRewriteArtifacts.PLAN_SCHEMA,
                file.toString(),
                "Application XML protection attributes cannot be inventoried safely",
                "Fix the malformed resource XML before Release",
            ).failure()
        }
    }

    private fun secureFactory(): DocumentBuilderFactory {
        val factory = DocumentBuilderFactory.newInstance()
        factory.isNamespaceAware = true
        factory.setFeature("http://apache.org/xml/features/disallow-doctype-decl", true)
        factory.setFeature("http://xml.org/sax/features/external-general-entities", false)
        factory.setFeature("http://xml.org/sax/features/external-parameter-entities", false)
        factory.setAttribute(XMLConstants.ACCESS_EXTERNAL_DTD, "")
        factory.setAttribute(XMLConstants.ACCESS_EXTERNAL_SCHEMA, "")
        return factory
    }

    private fun glob(selector: String): Pattern {
        val regex = StringBuilder("^")
        for (character in selector) {
            when (character) {
                '*' -> regex.append(".*")
                '?' -> regex.append('.')
                else -> regex.append(Pattern.quote(character.toString()))
            }
        }
        return Pattern.compile(regex.append('$').toString())
    }

    private fun valuesType(tag: String, declaredType: String): String? {
        if (tag == "item" || tag == "public") {
            return declaredType.ifBlank { null }
        }
        if (tag == "string-array" || tag == "integer-array") return "array"
        if (tag == "declare-styleable") return "styleable"
        if (tag in setOf("eat-comment", "skip", "overlayable")) return null
        return tag
    }

    private fun resourceFileName(fileName: String): String {
        val ninePatch = fileName.indexOf(".9.")
        if (ninePatch > 0) return fileName.substring(0, ninePatch)
        val dot = fileName.indexOf('.')
        return if (dot > 0) fileName.substring(0, dot) else fileName
    }

    class Result(
        @get:JvmName("resources") val resources: Set<BundleRewriteModule.ResourceKey>,
        @get:JvmName("protectedNames") val protectedNames: Set<String>,
        warnings: List<String>,
    ) {
        @get:JvmName("warnings")
        val warnings: List<String> = warnings.sorted()
    }
}
